src/jit/morph.cpp

```diff
@@ -8245,6 +8245,47 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
                                                      (call->gtCallType == CT_USER_FUNC) ? call->gtCallMethHnd : nullptr,
                                                      isTailPrefixed, TAILCALL_FAIL, szFailReason);
 
+#if FEATURE_MULTIREG_RET
+            if (fgGlobalMorph && call->HasMultiRegRetVal())
+            {
+                // The tail call has been rejected so we must finish the work deferred
+                // by impFixupCallStructReturn for multi-reg-returning calls and transform
+                //     ret call
+                // into
+                //     temp = call
+                //     ret temp
+
+                // Create a new temp.
+                unsigned tmpNum =
+                    lvaGrabTemp(false DEBUGARG("Return value temp for multi-reg return (rejected tail call)."));
+                const bool unsafeValueClsCheck = false;
+                lvaSetStruct(tmpNum, call->gtRetClsHnd, unsafeValueClsCheck);
+                lvaTable[tmpNum].lvIsMultiRegRet = true;
+
+                // Create the assignment tree and morph it.
+                var_types structType = lvaTable[tmpNum].lvType;
+                GenTree*  dst        = gtNewLclvNode(tmpNum, structType);
```

Comment:
>> Can we reuse code from impAssignMultiRegTypeToVar to do that?

I didn't do it for two reasons:
1. `impAssignMultiRegTypeToVar` calls `impAssignTempGen`, which changes the type of the rhs here:
https://github.com/dotnet/coreclr/blob/4070994640bcc2c4c138b6f695d3fce14ea3efe7/src/jit/importer.cpp#L807
We don't want that to happen, it may undo what we've done here:
https://github.com/dotnet/coreclr/blob/4070994640bcc2c4c138b6f695d3fce14ea3efe7/src/jit/morph.cpp#L7985
For the test in this PR it would set the type from TYP_STRUCT back to TYP_SIMD16. Multi-reg-returning calls have to be typed TYP_STRUCT.

2. impAssignMultiRegTypeToVar is too general and not very efficient. In this case it would cause the following calls: `impAssignMultiRegTypeToVar` --> `impAssignTempGen` --> `impAssignStruct` --> `impAssignStructPtr` where `impAssignStruct` would create `GT_ADDR` node for the destination and  `impAssignStructPtr` would discard it and just use the operand of `GT_ADDR`.

>> impFixupCallStructReturn would mark ret (dst in this case) as GTF_DONT_CSE, when this method will mark only result with this flag

You probably meant `impAssignMultiRegTypeToVar`? No, `ret` in `impAssignMultiRegTypeToVar` corresponds to `result` in this method so there is no diff.

Reply:
Thank you.